flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionGraphFactory.java

```diff
@@ -203,10 +204,15 @@ private void tryRestoreExecutionGraphFromSavepoint(
             final CheckpointCoordinator checkpointCoordinator =
                     executionGraphToRestore.getCheckpointCoordinator();
             if (checkpointCoordinator != null) {
+                Optional<String> stateBackendName = executionGraphToRestore.getStateBackendName();
+                boolean changelogEnabled =
+                        stateBackendName.isPresent()
+                                && "ChangelogStateBackend".equals(stateBackendName.get());
```

Comment:
I'm not sure whether it's always possible to infer whether the changelog is enabled or not from `executionGraphToRestore.getStateBackendName()`.
In particular, will this work if `flink-conf.yaml` contains `state.backend.changelog.enabled=true`?

Reply:
> will this work if flink-conf.yaml contains state.backend.changelog.enabled=true and job doesn't specify changelog or even any backend?

Yes, it works from the test results.
 I set `state.backend.changelog.enabled: true`in flink-conf.yaml and run examples/SocketWindowWordCount.jar, here is JM log, we can see that changelog is enabled.
```
2022-04-20 18:52:46,580 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Initializing job 'Socket Window WordCount' (23bf9f9518357ade7bec63df910fae98).
2022-04-20 18:52:46,585 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Using restart back off time strategy NoRestartBackoffTimeStrategy for Socket Window WordCount (23bf9f9518357ade7bec63df910fae98).
2022-04-20 18:52:46,595 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Running initialization on master for job Socket Window WordCount (23bf9f9518357ade7bec63df910fae98).
2022-04-20 18:52:46,595 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Successfully ran initialization on master in 0 ms.
2022-04-20 18:52:46,600 INFO  org.apache.flink.runtime.scheduler.adapter.DefaultExecutionTopology [] - Built 1 new pipelined regions in 0 ms, total 1 pipelined regions currently.
2022-04-20 18:52:46,600 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - No state backend has been configured, using default (HashMap) org.apache.flink.runtime.state.hashmap.HashMapStateBackend@37e3b93e
2022-04-20 18:52:46,612 INFO  org.apache.flink.state.changelog.ChangelogStateBackend       [] - ChangelogStateBackend is used, delegating HashMapStateBackend.
```
Meanwhile, we can see that the `StateBackendName` is set in `DefaultExecutionGraph.enableCheckpointing()`, and `enableCheckpoint()` is called by `DefaultExecutionGraphBuilder.buildGraph()`  in https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphBuilder.java#L307

So, before we call `executionGraphToRestore.getStateBackendName()`, the  StateBackendName has been set. https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionGraphFactory.java#L149

Reply:
Thanks for verifying. Indeed, the whole config is currently available to the JM.

However, I'm not sure whether we should rely on it because:
1. TMs can still have different configuration (I'm not sure whether it's a theoretical situation or not).
2. Configuration is called `jobMasterConfiguration`; which means it doesn't have to be the whole configuration.

So far we tried to not rely on it IIRC.

We probably don't need to make this decision now, before deciding which approach to take.

Reply:
@rkhachatryan , `StateBackendLoader` is called both on JM and TM side. I think we don't need to worry about this.

Reply:
I meant that there is no guarantee that JM and TM call `StateBackendLoader` with the same configuration.

Reply:
Though we cannot guarantee that the `flink-conf` same both on JM and TM side, however, this usually does not happen in production environment. What's more, unmatched configurations would also bring other problems except this.
I think we don't need to focus on this cocern case, we should focus on the general prons and cons of which each solution would make.

Reply:
I totally agree, that's what I mean above by "We probably don't need to make this decision now, before deciding which approach to take."